### ISSUE TITLE
[Feature Request]: OAuth Support for ATProtoKit

### ISSUE THREAD

**Issue by MasterJ93** [MasterJ93/ATProtoKit#43]
### Summary

OAuth support will need to be implemented for authenticating through the AT Protocol services.

### Pain points

App Passwords will be eventually deprecated. ATProtoKit needs to get ahead of this by implementing OAuth, which will be the new way for users to log into Bluesky and other services that are built on the AT Protocol.

### Is this a breaking change?

No

### Library Examples

The [Rust API library](https://github.com/sugyan/atrium/tree/main/atrium-oauth) has implemented OAuth from what I can gather. We can possibly convert the Rust code into Swift code.

### Additional Context

This is a fairly difficult part of the AT Protocol to understand. Even if I look at the Rust API library linked above and convert it into Swift code, it's still going to be challenging to understand.

That said, for future reference, these are some links that may help:
- [Bluesky Doc's webpage for OAuth](https://docs.bsky.app/docs/advanced-guides/oauth-client).
- [AT Protocol Doc's webpage for OAuth](https://atproto.com/specs/oauth).
- [The original design specification proposal](https://github.com/bluesky-social/proposals/tree/main/0004-oauth)

* The last one is not authoritative, but it may help.

**Comment by tonyarnold** [comment]
You should consider using a third party library like https://github.com/ChimeHQ/OAuthenticator to get the initial implementation in place, then decide if you need to build this out yourself.

**Comment by MasterJ93** [comment]
I'll look into that library and see if it works well for my needs. But I fear that I will have to write it by scratch after all (or at least grab a few projects and put them together).

**Comment by aaronvegh** [comment]
With social cards now available, this is my next-most-wanted feature. :-)

**Comment by knotbin** [comment]
+1 this is my biggest feature request

**Comment by MasterJ93** [comment]
I understand this and I'm working on trying to figure out a solution. I'll soon release a branch of my progress with this and I hope that others will be able to help contribute code and other ideas into this, since this is a top priority for me as well.

I just want to do this right: OAuth is a fairly complicated thing to understand.

**Comment by knotbin** [comment]
Of course! Completely fine to take your time, thank you so much for all the work you're doing on this.

**Comment by MasterJ93** [comment]
I'm well aware of that and I plan on having a tutorial of implementing that once I get the session configuration rework made. Please note that that repository requires Swift 5.10 or later.

**Comment by aaronvegh** [comment]
Thanks Christopher. Swift 5.10 won't be an impediment to my implementation; I'm looking forward to seeing what you've got!

**Comment by MasterJ93** [comment]
I'll _try_ to get it done by this week. You've waited long enough.

**Comment by knotbin** [comment]
Any progress on this?

**Comment by MasterJ93** [comment]
This part of the project is more or less going to be a summer project, as there's a ton of moving parts that I need to implement before I can properly work on this.

**Comment by aaronvegh** [comment]
Hey @MasterJ93 just sending my warmest wishes for this project. As I've said before, Bluesky auth remains the number one problem for users of my Bluesky app. There've been a rash of complaints lately so I just wanted to check in and see if work is moving forward on this? 🙏

**Comment by MasterJ93** [comment]
This is an extremely complicated part of the AT Protocol, so while I have laid the groundwork on it, it's been extremely slow. This is on top of all of the other AT Protocol-related projects I've been working on. That's not an excuse but rather an explanation, and so I apologize for that; I'll try my best to get OAuth support ready as soon as possible, but it will take some time to not only have it ready, but easy enough to just plug into ATProtoKit.

However, I believe ATProtoKit should have enough hooks to use something like OAuthenticator to add in OAuth support into ATProtoKit. The instructions are a bit complicated for me right now, but hopefully it won't be for you.

I should make a word of caution: once you do get it to work, I have to say that the simplest version doesn't keep users logged in indefinitely. That is, it keeps users logged it for a few weeks at a time, forcing users to re-login. I know that this is a sore spot for users of other third-party applications so please keep that in mind. There's a way to make this longer but it either requires the user to either use their own PDS or for you to have a dedicated server: both of which may not be ideal. This isn't a limitation of ATProtoKit, but of the AT Protocol itself. Once I get my OAuth package ready, I will try my best to document this.

**Comment by aaronvegh** [comment]
HI Chris, thanks for the update. I appreciate the difficulty of implementing OAuth, which I say without any credible expertise in the protocol at all. I'm not going to try pinning you down on a timeframe, and I don't even think I can say anything about other AT Protocol projects you've been working on, because none of that stuff means anything to me. But without a reliable authentication mechanism, I am in danger of losing users. I'm sure I have lost many who never write to me to complain.

I'm curious about your statement about using OAuthenticator. You're suggesting I use that library to implement the solution against ATProtoKit myself? It sounds like from what you're saying, you've tried it too but found the shelf life of the login was limited? I wonder if you have any additional information that you could share?

**Comment by MasterJ93** [comment]
[ATOAuthKit's repository is live today.](https://github.com/ATProtoKit/ATOAuthKit)

At the moment, the only child package is the OAuthTypes package. Once the main ATOAuthKit package (which would include the OAuth client) has been implemented (with documentation), this issue will officially be closed.

The goal is to complete the client side before the end of the month.

**Comment by MasterJ93** [comment]
The `oauth` branch of ATProtoKit is now available and has been merged to the [`develop`](https://github.com/MasterJ93/ATProtoKit/tree/develop) branch.

There are a number of changes to it, so you can look at some of them in the [Pull Request](https://github.com/MasterJ93/ATProtoKit/pull/310).

It goes without saying that assistance is needed for this package, but with respect to OAuth in particular, I will need all the assistance I can get. I feel like there's still a ton of rough edges to smooth out, as well as some inconsistencies, so I appreciate it if you could help me out with that.

A DocC article with tutorials on how to use third-party OAuth packages will be available soon. When ATOAuthKit is updated with its own client implementation, the article will be updated, though keep in mind that ATOAuthKit will need more time to be worked on.

**Comment by MasterJ93** [comment]
The `oauth` branch of ATProtoKit has been merged with the `develop` branch, and that branch has been merged to `main`.

A walkthrough for creating a macOS app using OAuth has been added as well.

I will do my best to release version 0.34.0 of ATProtoKit on Wednesday; the only thing left to do is to compile the release notes.

**Comment by MasterJ93** [comment]
Version 0.34.0 of ATProtoKit is out now with OAuth support.

As a result, I will be unpinning this issue, but I will _not_ be closing this yet until:
- The tutorials are easy enough to understand.
- DocC tutorials and video tutorials are out.
- All aspects of OAuth are documented.
- ATOAuthKit is implemented enough to make it extremely easy for developers to integrate OAuth with just a few lines of code.

Thank you all for waiting so patiently for this. I appreciate it immensely.
